src/main/java/io/javalin/core/JavalinServlet.kt

```diff
@@ -107,26 +106,36 @@ class JavalinServlet(
                     }
                     tryErrorHandlers()
                     tryAfterHandlers()
-                    writeResult(ctx, asyncContext.response as HttpServletResponse)
+                    writeResult(ctx, asyncContext.response as HttpServletResponse, type)
+                    logRequest(ctx)
                     asyncContext.complete()
                 }
             }
         }
     }
 
-    private fun writeResult(ctx: Context, res: HttpServletResponse) {
-        if (!res.isCommitted) {
-            ctx.resultStream()?.let { resultStream ->
-                if (gzipShouldBeDone(ctx)) {
-                    GZIPOutputStream(res.outputStream, true).use { gzippedStream ->
-                        res.setHeader(Header.CONTENT_ENCODING, "gzip")
-                        resultStream.copyTo(gzippedStream)
-                    }
-                } else {
-                    resultStream.copyTo(res.outputStream)
-                }
+    private fun writeResult(ctx: Context, res: HttpServletResponse, type: HandlerType) {
+        if (res.isCommitted || ctx.resultStream() == null) return // nothing to write
+        val resultStream = ctx.resultStream()!!
+        if (res.getHeader(Header.ETAG) != null || (autogeneratedEtagsEnabled && type == HandlerType.GET)) {
+            val serverEtag = res.getHeader(Header.ETAG) ?: Util.getChecksumAndReset(resultStream) // calculate if not set
```

Comment:
this could perhaps be it's own method where you return true if you should return with a 304?